libobs/obs-output.c

```diff
@@ -2699,3 +2699,34 @@ const char *obs_output_get_protocols(const obs_output_t *output)
 		       ? output->info.protocols
 		       : NULL;
 }
+
+bool obs_enum_output_types_with_protocol(const char *protocol, size_t idx,
+					 const char **id)
+{
+	if (!obs_is_output_protocol_registered(protocol))
+		return false;
+
+	size_t count = 0;
+	size_t protocol_len = strlen(protocol);
+	for (size_t i = 0; i < obs->output_types.num; i++) {
+		if (!(obs->output_types.array[i].flags & OBS_OUTPUT_SERVICE))
+			continue;
+
+		const char *substr = obs->output_types.array[i].protocols;
+		while (substr && substr[0] != '\0') {
+			const char *next = strchr(substr, ';');
+			size_t len = next ? (size_t)(next - substr)
+					  : strlen(substr);
+			if (protocol_len == len &&
+			    strncmp(substr, protocol, len) == 0) {
+				if (count == idx) {
+					*id = obs->output_types.array[i].id;
+					return true;
+				}
+				count++;
+			}
+			substr = next ? next + 1 : NULL;
+		}
+	}
+	return false;
+}
```

Comment:
This enum has future purpose/usage in RFC 39, I put it in RFC 45 to avoid creating something like `obs_get_first_output_type_with_protocol` just for a transition period.

I will check the `enum_cb` suggestion.